### PR TITLE
Added notifications package to unmarshal webhook notifications (BLAPI-37)

### PR DESCRIPTION
**Pull request by katejefferson**
Added notifications package and parse method with tests. The caller (wherever the incoming webhook is received) will need to make an API call to Recurly to `LookupInvoice` to confirm the validity of the webhook before sending a message to the subscription service.

### CODE REVIEW

File: recurly/notifications/xml/failed_payment_notification.xml [hunk 1]
```diff
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (cristiangraz) [inline recurly/notifications/xml/failed_payment_notification.xml]: Can you rename the `xml` folder to `testdata`?

> Directory and file names that begin with "." or "_" are ignored by the go tool, as are directories named "testdata".

File: recurly/notifications/notifications_test.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+	"os"
+	"reflect"
+	"testing"
+
+	"github.com/blacklightcms/go-recurly/recurly"
+	"github.com/blacklightcms/go-recurly/recurly/notifications"
+)
+
+func TestParse_SuccessfulPaymentNotification(t *testing.T) {
+	xmlFile, err := os.Open("xml/successful_payment_notification.xml")
```
Comment (cristiangraz) [inline recurly/notifications/notifications_test.go]: Can you create a `MustOpen` test function that panics on err?

File: recurly/notifications/notifications_test.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer xmlFile.Close()
```
Comment (cristiangraz) [inline recurly/notifications/notifications_test.go]: You should be able to remove this, `xmlFile` implements `io.Closer`, which the `Parse` function is closing.

File: recurly/notifications/notifications_test.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+
+	req, err := http.NewRequest("POST", "/", xmlFile)
```
Comment (cristiangraz) [inline recurly/notifications/notifications_test.go]: `os.Open` returns `*os.File`, which [implements](https://golang.org/pkg/os/#File.Read) the `io.Reader` interface. You can pass `xmlFile` directly into `notifications.Parse` without the extra step of creating an `*http.Request`.

File: recurly/notifications/notifications_test.go [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+	switch result.(type) {
+	case *notifications.SuccessfulPaymentNotification:
+		n, ok = result.(*notifications.SuccessfulPaymentNotification)
+		if !ok {
+			t.Fatalf("unable to reflect interface")
+		}
+		break
+	default:
+		t.Fatalf("unexpected result type: %T", result)
+	}
```
Comment (cristiangraz) [inline recurly/notifications/notifications_test.go]: Can you remove the switch statement? The inner type assertion is all you need.

```go
if n, ok := result.(*notifications.SuccessfulPaymentNotification); !ok {
   t.Fatal("unexpected type: %T", result)
} else if !reflect.DeepEqual(n, &notifications.SuccessfulPaymentNotification{...}) {
   t.Fatal(...)
}
```

File: recurly/notifications/notifications.go [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+	}
+
+	// PastDueInvoiceNotification is sent when an invoice is past due.
+	PastDueInvoiceNotification struct {
+		Account recurly.Account
+		Invoice recurly.Invoice
+	}
+)
+
+// TODO respond to webhook with a 200 code within 5 minutes or Recurly will resend it.
```
Comment (cristiangraz) [inline recurly/notifications/notifications.go]: Can this block of comments be removed?

File: recurly/notifications/xml/successful_payment_notification.xml [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+  <account>
+    <account_code>1</account_code>
+    <username nil="true">verena</username>
+    <email>verena@example.com</email>
+    <first_name>Verena</first_name>
+    <last_name>Example</last_name>
+    <company_name nil="true">Company, Inc.</company_name>
+  </account>
+  <transaction>
+    <id>a5143c1d3a6f4a8287d0e2cc1d4c0427</id>
```
Comment (cristiangraz) [inline recurly/notifications/xml/successful_payment_notification.xml]: It looks like this field may be getting discarded (for all notifications with a transaction); I don't see this value referenced in any of the `reflect.DeepEqual` assertions.

File: recurly/notifications/notifications.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+
+	var dst interface{}
+	switch n.XMLName.Local {
+	case SuccessfulPayment:
+		dst = &SuccessfulPaymentNotification{}
+	case FailedPayment:
+		dst = &FailedPaymentNotification{}
+	case PastDueInvoice:
+		dst = &PastDueInvoiceNotification{}
+	}
```
Comment (cristiangraz) [inline recurly/notifications/notifications.go]: Can you add a `default` case to this switch statement that returns `nil, ErrUnknownNotification` where `ErrUnknownNotification` is a defined package var? The `ErrUnknownNotification` should be a type that implements the `error` interface and returns the name of the notification in the error string as well as having a `Name()` method that returns the notification name from the XML.

That will allow the caller to know if the error wasn't in parsing the notification, but specifically that the notification doesn't exist or wasn't one of the defined types.

File: recurly/transactions.go [hunk 9]
```diff
@@ -17,6 +17,7 @@ type (
 	Transaction struct {
 		XMLName         xml.Name  `xml:"transaction"`
 		Invoice         href      `xml:"invoice,omitempty"`
+		InvoiceNumber   int       `xml:"invoice_number,omitempty"`
```
Comment (cristiangraz) [inline recurly/transactions.go]: Is this part of just the webhooks? if so, could you remove from here and put just in the webhooks package? API users might expect that value to be populated if the field is on the struct.

File: recurly/notifications/notifications.go (hunk 6 above)
Comment (katejefferson) [inline recurly/notifications/notifications.go]: Yes, I just wanted to ensure that somewhere it is noted that webhooks need to receive a response from our implementation.

File: recurly/notifications/notifications_test.go (hunk 3 above)
Comment (katejefferson) [inline recurly/notifications/notifications_test.go]: Removed in [910c54](https://github.com/blacklightcms/go-recurly/pull/6/commits/15fba61ea259a3c8441c3a2f23a20c5319910c54)

File: recurly/notifications/notifications_test.go (hunk 4 above)
Comment (katejefferson) [inline recurly/notifications/notifications_test.go]: Removed in [910c54](https://github.com/blacklightcms/go-recurly/pull/6/commits/15fba61ea259a3c8441c3a2f23a20c5319910c54)

File: recurly/notifications/notifications_test.go (hunk 5 above)
Comment (katejefferson) [inline recurly/notifications/notifications_test.go]: Removed in [910c54](https://github.com/blacklightcms/go-recurly/pull/6/commits/15fba61ea259a3c8441c3a2f23a20c5319910c54)

File: recurly/notifications/notifications.go (hunk 6 above)
Comment (katejefferson) [inline recurly/notifications/notifications.go]: Removed in [910c54](https://github.com/blacklightcms/go-recurly/pull/6/commits/15fba61ea259a3c8441c3a2f23a20c5319910c54)

File: recurly/notifications/notifications.go (hunk 8 above)
Comment (katejefferson) [inline recurly/notifications/notifications.go]: Added in [910c54](https://github.com/blacklightcms/go-recurly/pull/6/commits/15fba61ea259a3c8441c3a2f23a20c5319910c54)

File: recurly/notifications/xml/successful_payment_notification.xml (hunk 7 above)
Comment (katejefferson) [inline recurly/notifications/xml/successful_payment_notification.xml]: Updated in [910c54](https://github.com/blacklightcms/go-recurly/pull/6/commits/15fba61ea259a3c8441c3a2f23a20c5319910c54)

File: recurly/notifications/notifications.go [hunk 10]
```diff
@@ -0,0 +1,132 @@
+package notifications
```
Comment (cristiangraz) [inline recurly/notifications/notifications.go]: Could you rename `webhooks`?

File: recurly/notifications/notifications.go [hunk 11]
```diff
@@ -0,0 +1,132 @@
...
+
+// transactionHolder allows the uuid and invoice number fields to be set.
+// The UUID field is labeled id in notifications and the invoice number
+// is not included on the existing transaction struct.
+type transactionHolder interface {
+	setTransactionFields(id string, in int)
+}
+
+// setTransactionFields sets fields on the transaction struct.
+func (n *SuccessfulPaymentNotification) setTransactionFields(id string, in int) {
```
Comment (cristiangraz) [inline recurly/notifications/notifications.go]: Could you rename `in` to `invoiceNumber`?

File: recurly/notifications/notifications.go [hunk 12]
```diff
@@ -0,0 +1,132 @@
...
+	n.Transaction.UUID = id
+	n.InvoiceNumber = in
```
Comment (cristiangraz) [inline recurly/notifications/notifications.go]: With the latest version on `master`, you can now set `n.Transaction.InvoiceNumber` and remove the `InvoiceNumber` from each of the `Notification` structs.

Comment (katejefferson) [inline recurly/notifications/notifications.go]: Updated in [9f2146](https://github.com/blacklightcms/go-recurly/pull/6/commits/ded5ee95deb68ddc6194b97100ac475f579f2146)

File: recurly/notifications/notifications.go (hunk 11 above)
Comment (katejefferson) [inline recurly/notifications/notifications.go]: Updated in [9f2146](https://github.com/blacklightcms/go-recurly/pull/6/commits/ded5ee95deb68ddc6194b97100ac475f579f2146)

File: recurly/notifications/notifications.go (hunk 10 above)
Comment (katejefferson) [inline recurly/notifications/notifications.go]: Updated in [9f2146](https://github.com/blacklightcms/go-recurly/pull/6/commits/ded5ee95deb68ddc6194b97100ac475f579f2146)

**Pull request merged**